### PR TITLE
[2.x] Revamp the performance-related website content

### LINKED ISSUE #2533: Revamp the `Performance` page
As agreed in [the site structure document](https://docs.google.com/document/d/10Fu7oqDzdM_D6LbexzwX9arh51Tic7AGvkWTQrL6jjQ/edit), revamp the `Performance` page such that

- Outline:
  - Asynchronous logging
  - Filtering
  - Source location information
  - Memory management (i.e., garbage-free & recyclers)
    - Add a `Performance` section to the `Pattern Layout` page and link to it
    - Add a `Performance` section to the `JSON Template Layout` page and link to it
- Revamp `garbagefree.adoc`
- Target `2.x`, port to `main`

### PR DESCRIPTION
**Pull request by vy**
Fixes #2533.

### Reviewer notes

* Reviewers are recommended to checkout the branch locally, build, and view it.
* Except trivial typo/grammar fixes, **do not push content changes to the branch**, instead drop a comment to the associated parts of the source, please.

### CODE REVIEW

File: src/site/antora/modules/ROOT/pages/manual/layouts.adoc [hunk 1]
```diff
@@ -16,39 +16,25 @@
 ////
 = Layouts
 
-An Appender uses a Layout to format a LogEvent into a form that meets the needs of whatever will be consuming the log event.
-In Log4j 1.x and Logback Layouts were expected to transform an event into a String.
-In Log4j 2 Layouts return a byte array.
-This allows the result of the Layout to be useful in many more types of Appenders.
-However, this means you need to configure most Layouts with a
+An Appender uses a Layout to format a LogEvent into a form that meets
```
Comment (ppkarwasz) [inline src/site/antora/modules/ROOT/pages/manual/layouts.adoc:19]: Are there any changes in this file?

It seems it was only accidentally reformatted.

File: src/site/antora/modules/ROOT/pages/manual/garbagefree.adoc [hunk 2]
```diff
@@ -14,565 +14,230 @@
...
+(See xref:manual/configuration.adoc[] on details how you can set these properties.)
 
-|%t, %tn, %thread, %threadName
-|The name of the thread that generated
-the logging event
+include::partial$manual/systemproperties/properties-garbage-collection.adoc[leveloffset=+2,tag=api]
 
-|%tp
-|The priority of the thread that generated the logging event
+include::partial$manual/systemproperties/properties-thread-context.adoc[leveloffset=+2,tag=gcfree]
```
Comment (ppkarwasz) [inline src/site/antora/modules/ROOT/pages/manual/garbagefree.adoc:170]: As in the remark above: the algorithm that interprets which `ThreadContextMap` implementation to use has been moved from the `ThreadContextMapFactory` class (in `log4j-api`) to implementations of `Provider` (in `log4j-core` and the other Log4j API implementations).

That is why I would consider this part of Log4j Core configuration.

File: src/site/antora/modules/ROOT/nav.adoc [hunk 3]
```diff
@@ -52,14 +52,21 @@
...
-** xref::manual/customconfig.adoc[]
+** xref:manual/customconfig.adoc[]
 * xref:manual/plugins.adoc[]
+* xref:manual/customconfig.adoc[]
+* xref:manual/customloglevels.adoc[]
 * xref:manual/jmx.adoc[]
 * xref:manual/logsep.adoc[]
 * xref:manual/performance.adoc[]
+** xref:manual/async.adoc[]
+** xref:manual/garbagefree.adoc[]
```
Comment (ppkarwasz) [inline src/site/antora/modules/ROOT/nav.adoc:69]: I find this kind of hierarchy a little bit weird.

I would expect to see a short text under "Performance" and 3 children: "Log4j API usage", "Asynchronous loggers" and "Garbage Free".

**Review by ppkarwasz — APPROVED** [review]
Looks mostly good to me.

File: src/site/antora/modules/ROOT/partials/manual/systemproperties/properties-garbage-collection.adoc [hunk 4]
```diff
@@ -56,6 +56,8 @@ The size in ``char``s of the link:../https://docs.oracle.com/javase/8/docs/api/j
 
 This setting is only used if <<log4j2.enableDirectEncoders>> is set to `true`.
 
+// tag::api[]
+
 [id=log4j2.initialReusableMsgSize]
 == `log4j2.initialReusableMsgSize`
```
[the remark this answers is not in the document]
Comment (vy) [reply src/site/antora/modules/ROOT/partials/manual/systemproperties/properties-garbage-collection.adoc]: In ea573b384d800a076b45e5c7e21d6b00eadc59fc,

1. Added a note at the top of `garbagefree.adoc` and stated garbage-free only applies to Log4j API and Log4j Core
2. Moved all shared properties to the `Log4j Core configuration > Properties` section

File: src/site/antora/modules/ROOT/pages/manual/garbagefree.adoc (hunk 2 above)
Comment (vy) [reply src/site/antora/modules/ROOT/pages/manual/garbagefree.adoc]: Moved all shared properties to the `Log4j Core configuration > Properties` section in ea573b384d800a076b45e5c7e21d6b00eadc59fc.

File: src/site/antora/modules/ROOT/nav.adoc (hunk 3 above)
Comment (vy) [reply src/site/antora/modules/ROOT/nav.adoc:69]: As can be seen in `performance.adoc`, it is not just about Log4j API, async. loggers, and garbage-free. There are more concerns that need to be covered. I think `performance.adoc` is in a good state now:

* It is a _complete_ portal to all performance-related knobs
* The content is ordered from generic to specific

My rationale for keeping async. loggers and garbage-free in separate pages is:

* They are optional and advanced features
* Their setup spans across both API and implementation (unfit for the easy-to-follow procedural checklist established in `performance.adoc`)
* Their content is massive
* Link compatibility concerns

**Pull request merged**